stan/math/rev/functor/reduce_sum.hpp

```diff
@@ -89,55 +108,93 @@ struct reduce_sum_impl<ReduceFunction, require_var_t<ReturnType>, ReturnType,
         return;
       }
 
-      if (args_adjoints_.size() == 0) {
-        args_adjoints_ = Eigen::VectorXd::Zero(num_vars_shared_terms_);
-      }
-
       // Initialize nested autodiff stack
       const nested_rev_autodiff begin_nest;
 
-      // Create nested autodiff copies of sliced argument that do not point
-      //   back to main autodiff stack
-      std::decay_t<Vec> local_sub_slice;
-      local_sub_slice.reserve(r.size());
-      for (size_t i = r.begin(); i < r.end(); ++i) {
-        local_sub_slice.emplace_back(deep_copy_vars(vmapped_[i]));
+      // Create copies of sliced argument which refer to the main
+      // thread
+      std::decay_t<Vec> local_sub_slice(vmapped_.begin() + r.begin(),
+                                        vmapped_.begin() + r.end());
+
+      // first save current adjoints of sliced argument (which we
+      // restore later) and set these to zero
+      Eigen::VectorXd adjoints_sub_slice(r.size() * num_vars_per_term_);
```

Comment:
Can this variable be like `saved_adjoints`/`adjoints_backup` or something to indicate it is being used to save stuff in the forward pass as opposed to stuff for the reverse pass?